www/updateUes.php

```diff
@@ -6,20 +6,122 @@
  * @author Pavel Vyskočil <vyskocilpavel@muni.cz>
  */
 
+use Jose\Component\Checker\ClaimCheckerManager;
+use Jose\Component\Checker;
+use Jose\Component\Core\AlgorithmManager;
+use Jose\Component\KeyManagement\JWKFactory;
+use Jose\Component\Signature\Algorithm\RS512;
+use Jose\Component\Signature\JWSVerifier;
+use Jose\Component\Checker\HeaderCheckerManager;
+use Jose\Component\Checker\AlgorithmChecker;
+use Jose\Component\Signature\JWSTokenSupport;
+use Jose\Component\Signature\Serializer\CompactSerializer;
+use Jose\Component\Signature\Serializer\JWSSerializerManager;
+use SimpleSAML\Configuration;
 use SimpleSAML\Logger;
 use SimpleSAML\Module\perun\Adapter;
+use SimpleSAML\Module\perun\DatabaseConnector;
 
 $adapter = Adapter::getInstance(Adapter::RPC);
 
 $entityBody = file_get_contents('php://input');
-$body = json_decode($entityBody, true);
+$token = json_decode($entityBody, true);
 
-$attributesFromIdP = $body['attributes'];
-$attrMap = $body['attrMap'];
-$attrsToConversion = $body['attrsToConversion'];
-$perunUserId = $body['perunUserId'];
+$attributesFromIdP = null;
+$attrMap = null;
+$attrsToConversion = null;
+$perunUserId = null;
+$id = null;
 
 const UES_ATTR_NMS = 'urn:perun:ues:attribute-def:def';
+const TABLE_NAME = 'scriptChallenges';
+const CHALLENGE = 'challenge';
+const CONFIG_FILE_NAME = 'keys.php';
+
+try {
+    $config = Configuration::getConfig(CONFIG_FILE_NAME);
+    $keyPub = $config->getString('updateUes');
+
+    $algorithmManager = new AlgorithmManager([new RS512()]);
+    $jwsVerifier = new JWSVerifier($algorithmManager);
+    $jwk = JWKFactory::createFromKeyFile($keyPub);
+
+    $serializerManager = new JWSSerializerManager([new CompactSerializer()]);
+    $jws = $serializerManager->unserialize($token);
+
+    $headerCheckerManager = new HeaderCheckerManager([new AlgorithmChecker(['RS512'])], [new JWSTokenSupport()]);
+    $headerCheckerManager->check($jws, 0);
+
+    $isVerified = $jwsVerifier->verifyWithKey($jws, $jwk, 0);
+
+    if (!$isVerified) {
+        die('The token signature is invalid!');
+    }
+
+    $claimCheckerManager = new ClaimCheckerManager(
+        [
+            new Checker\IssuedAtChecker(),
+            new Checker\NotBeforeChecker(),
+            new Checker\ExpirationTimeChecker(),
+        ]
+    );
+
+    $claims = json_decode($jws->getPayload(), true);
+    $claimCheckerManager->check($claims);
+
+    $challenge = $claims['challenge'];
+
+    $attributesFromIdP = $claims['data']['attributes'];
+    $attrMap = $claims['data']['attrMap'];
+    $attrsToConversion = $claims['data']['attrsToConversion'];
+    $perunUserId = $claims['data']['perunUserId'];
+    $id = $claims['id'];
+
+    $databaseConnector = new DatabaseConnector();
+    $conn = $databaseConnector->getConnection();
+
+    assert($conn !== null);
+
+    $stmt = $conn->prepare('SELECT challenge FROM ' . TABLE_NAME . ' WHERE id=?');
```

Comment:
the challenge should also have a limited lifetime, so either add SQL condition here to select challenges younger than e.g. 5 minutes, or add a cron job which deletes challenges older than e.g. 5 minutes (into the `hooks` folder, see [proxystatistics](https://github.com/CESNET/proxystatistics-simplesamlphp-module/blob/master/hooks/hook_cron.php) for inspiration)